File: qc-core/src/main/java/com/eager/questioncloud/core/domain/payment/point/model/ChargePointPayment.java
package com.eager.questioncloud.core.domain.payment.point.model;

import com.eager.questioncloud.core.domain.payment.point.vo.ChargePointPaymentStatus;
import com.eager.questioncloud.core.domain.payment.point.vo.ChargePointType;
import com.eager.questioncloud.core.domain.portone.dto.PortonePayment;
import com.eager.questioncloud.core.exception.CustomException;
import com.eager.questioncloud.core.exception.Error;
import com.eager.questioncloud.core.exception.InvalidPaymentException;
import java.time.LocalDateTime;
import lombok.Builder;
import lombok.Getter;

@Getter
public class ChargePointPayment {
    private final String paymentId;
    private final Long userId;
    private String receiptUrl;
    private final ChargePointType chargePointType;
    private ChargePointPaymentStatus chargePointPaymentStatus;
    private final LocalDateTime createdAt;
    private LocalDateTime paidAt;

    @Builder
    public ChargePointPayment(String paymentId, Long userId, String receiptUrl, ChargePointType chargePointType,
        ChargePointPaymentStatus chargePointPaymentStatus, LocalDateTime createdAt, LocalDateTime paidAt) {
        this.paymentId = paymentId;
        this.userId = userId;
        this.receiptUrl = receiptUrl;
        this.chargePointType = chargePointType;
        this.chargePointPaymentStatus = chargePointPaymentStatus;
        this.createdAt = createdAt;
        this.paidAt = paidAt;
    }


    public static ChargePointPayment order(String paymentId, Long userId, ChargePointType chargePointType) {
        return ChargePointPayment.builder()
            .paymentId(paymentId)
            .userId(userId)
            .chargePointType(chargePointType)
            .chargePointPaymentStatus(ChargePointPaymentStatus.ORDERED)
            .createdAt(LocalDateTime.now())
            .build();
    }

    public void approve(PortonePayment payment) {
        validatePayment(payment);
        this.chargePointPaymentStatus = ChargePointPaymentStatus.PAID;
        this.receiptUrl = payment.getReceiptUrl();
        this.paidAt = LocalDateTime.now();
    }

    public void fail() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.Fail;
    }

    private void validatePayment(PortonePayment payment) {
        validateStatus();
        validateAmount(payment);
    }

    private void validateAmount(PortonePayment payment) {
        if (chargePointType.getAmount() != payment.getAmount().getTotal()) {
            throw new InvalidPaymentException(payment);
        }
    }

    private void validateStatus() {
        if (!chargePointPaymentStatus.equals(ChargePointPaymentStatus.ORDERED)) {
            throw new CustomException(Error.ALREADY_PROCESSED_PAYMENT);
        }
    }
}

File: qc-domain/qc-point/src/main/kotlin/com/eager/questioncloud/point/domain/ChargePointPayment.kt
package com.eager.questioncloud.point.domain

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.point.enums.ChargePointPaymentStatus
import com.eager.questioncloud.point.enums.ChargePointType
import com.eager.questioncloud.point.exception.InvalidPointChargeException
import io.hypersistence.tsid.TSID
import java.time.LocalDateTime

class ChargePointPayment(
    val orderId: String = TSID.Factory.getTsid().toString(),
    var paymentId: String? = null,
    val userId: Long,
    val chargePointType: ChargePointType,
    var chargePointPaymentStatus: ChargePointPaymentStatus = ChargePointPaymentStatus.ORDERED,
    val createdAt: LocalDateTime = LocalDateTime.now(),
    var requestAt: LocalDateTime? = null,
) {
    fun prepare(paymentId: String) {
        this.paymentId = paymentId
        this.chargePointPaymentStatus = ChargePointPaymentStatus.PAYMENT_REQUEST
        this.requestAt = LocalDateTime.now()
    }

    fun charge() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.CHARGED
    }

    fun cancel() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.CANCELED
    }

    fun failCancel() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.PAYMENT_REQUEST
    }

    fun validatePayment(paidAmount: Int) {
        validateStatus()
        validateAmount(paidAmount)
    }

    private fun validateAmount(paidAmount: Int) {
        if (chargePointType.amount != paidAmount) {
            throw InvalidPointChargeException(paymentId!!)
        }
    }

    private fun validateStatus() {
        if (chargePointPaymentStatus != ChargePointPaymentStatus.ORDERED) {
            throw CoreException(Error.ALREADY_PROCESSED_PAYMENT)
        }
    }

    companion object {
        fun createOrder(
            userId: Long,
            chargePointType: ChargePointType
        ): ChargePointPayment {
            return ChargePointPayment(
                userId = userId,
                chargePointType = chargePointType
            )
        }
    }
}
